### ISSUE TITLE
AppWindow...

### ISSUE THREAD

**Issue by amwx** [amwx/FluentAvalonia#303]
Edit: this has been merged into master and will be available in the next preview (#307)

With Avalonia's public API being locked down the way it is, a new problem has arisen around `AppWindow`. Currently, the way `AppWindow` works on Win32 is a custom `IWindowImpl` that derives from `Avalonia.Win32.WindowImpl` so that I can keep all the normal windowing functions, but make the frame adjustments I need to. Well, this type was moved to `internal` in a recent PR which will break `AppWindow` moving forward. So yet another change needs to be made to handle this, one that is future proof and won't break pending other API visibility changes. 

I'm testing out using a simple WndProc hook via `GetWindowLongPtrW` which so far looks very promising and will likely be what I do. However, we have a few things to address now:

1. Activating `FullScreen` mode via `WindowState` won't work anymore. This is managed principle as the Win32 API doesn't understand "FullScreen", and I was already hacking my way around the upstream logic so it works as intended. I won't be able to do this anymore, so `AppWindow` will need another method to handle FullScreen. 
2. The issue where the window doesn't respect a size set in Xaml or code behind is back, solving this outside of the window impl is going to take some thought.
3. There's a possibility this will regress the issue where minimizing/maximizing the window and then restoring it will cause it to shrink in size...however, so far I can't repro that on Win10 or 11 - that might've been caused by something else that I'm not remembering right now.

There is, however, an upside to this. Not having a reference to Win32.WindowImpl means I no longer have to take a reference to `Avalonia.Desktop` package. This also means, having AppWindow reside in a separate package is no longer necessary. Therefore I am likely to return AppWindow back into the main package and deprecate that package. It will keep the namespace however. This has a couple benefits:
- One less package to worry about
- One less StyleInclude to worry about
- No need to have an app builder extension for `TaskDialog` to work in windowed mode as other controls in the main library can see the AppWindow type. 
- This is also important for apps that split projects up by target platform, something like [MainProject, MainProject.Desktop, MainProject.Android, etc]. (if you can read between the lines here, knowing that I'm reworking the sample app, you can figure out why this is especially important)

I hate that we're bouncing around so much with this custom window, but doing all of this should future proof it and we should be done after this.

**Comment by timunie** [comment]
I really wonder why we even need this special window. Iirc, you want to fix issues that the upstream Avalonia window has. If that is the case, I wonder if fixing it upstream wouldn't be the better solution in this case?

**Comment by danielmayost** [comment]
@timunie i think it's not "fix issues", it's special new features such as custom title bar etc, actually it's style compatible with fluent v2.

**Comment by amwx** [comment]
> I really wonder why we even need this special window. Iirc, you want to fix issues that the upstream Avalonia window has. If that is the case, I wonder if fixing it upstream wouldn't be the better solution in this case?

The primary thing `AppWindow` (or its predecessor `CoreWindow`) set out to achieve is a win32 window with extended client area that also had the resize drag region OUTSIDE the window (in the drop shadow) like a normal window does. With the implementation of `DwmExtendFrameIntoClientArea` that's done upstream this isn't possible - the resize area is inside the client area of the window, which an lead to issues particularly with scrollbars. WPF's WindowChrome stuff works the exact same way. 

To fix this, there's an adjustment that's made when receiving the `WM_NCCALCSIZE` message that fixes this (which is what UWP an WinUI apps do under the hood), however, it has the side effect of not being able to remove the system window border, which will be undesirable to many people that do their own window chrome (I think even you've said something like this). So I don't deem something like this as something that should be handled upstream trying to manage which type of extended frame you want would be difficult (especially as this is Win32 only), and probably confusing.

There's also the recreation of the system-style titlebar that I do, which also includes ensuring the Win 11 snap layouts flyout shows when hovering the maximize button, and some of the new features I added alongside AppWindow. 

The main problem here is the way I was handling all this was relying on an API that was not always guaranteed to be public. To be fair though, `CoreWindow` was added ~2 years ago by request (it was originally only in the sample app) and when Avalonia wasn't nearly where it is today, this was the easiest way to do it, and there wasn't a whole lot that would indicate a rush to hide some of these types. Hindsight is always 20/20
